packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1837,6 +1838,7 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
           caretRect.left,
           caretRect.top - _kCaretHeightOffset,
           caretRect.width,
+          // TODO(bleroux): check if it is intended to use caretHeight instead of caretRect.height.
```

Comment:
My hunch is that it should be caretRect.height.  I can see [in _computeCaretPrototype](https://github.com/flutter/flutter/blob/445b08685a1acb2231f2d7d3fdc372eb547469c7/packages/flutter/lib/src/rendering/editable.dart#L2284) that caretRect.height is already adjusted from caretHeight.

Maybe @LongCatIsLooong would know more?

Reply:
While investigating this (and various text input issues), I too noticed that there is sometimes something wrong with the cursor and text alignment. I'm eager to see if https://github.com/flutter/flutter/issues/142969 will help.

> It seems weird to me that the line height is used as the branching threshold. The line height has no meaning without the font size being specified.

The main side of this issue is that the cursor is too tall on M3. The height of the cursor is directly correlated to the line height (because its default value is fontSize * lineHeight). On M2 this was ok because default line height was 1.0, on M3 default line height is 1.5, so the cursor is already very tall. 

Previously, on Apple platforms, an eye-ball value was added to the cursor height for fidelity, which made sense on M2. On M3, this increase does not make sense because the cursor is already tall enough (default font size is 16, line 1.5, the default cursor height is 24).

What this PR does is relying on line height to apply the previous existing increase or apply a small decrease otherwise. We can’t rely on font size to make this decision. Without this change, the default cursor height on M3 is `26` on Apple platforms and `20` on the other platforms. With this PR, it is `22` on Apple platforms (and still `20` on the other platforms).

This is a simple solution to get a better result for a default text field on M3 (without breaking existing M2 tests). As I commented inside the PR, a complete solution is needed (this is tracked on https://github.com/flutter/flutter/issues/120836 and https://github.com/flutter/flutter/issues/7034). Part of the long term solution would mean figuring out if the cursor height should be correlated to the line height.

Reply:
The current cursor positioning logic top-aligns the cursor with the line box (unless `enforceStrutHeight` is set to false). No matter how tall the cursor is, it shouldn't have a smaller y offset than the RenderEditable. I'd say let's wait for the fix for #142969 and see if the cursor still overlaps the label.

Reply:
I don't think those fixes are mutually exclusive:
- a fix for https://github.com/flutter/flutter/issues/142969 might avoid the label overlaping.
- this one is mitigating the previous eye-ball logic which extends the cursor on macOS/iOS.

Reply:
Sorry for the late reply. I'm hoping the alignment fix will land today/tomorrow. 

> this one is mitigating the previous eye-ball logic which extends the cursor on macOS/iOS.

From the comment it seems the new cursor height is still an eye-balled value (and it adds more magic to the previous value). On iOS the cursor height roughly equals the line height so `preferredLineHeight + 2` seems to be reasonable (although I have no idea why the "+2)?

Reply:
After taking a closer look at the cursor code, it does seem that on iOS and macOS the cursor's origin is usually about 1 pixel higher than the text box because `caretProtoType` is set to `cursorHeight + 2`.  But I'm still skeptical about using the `height` multiplier to determine the default height of the cursor.  It's introducing heuristics and I think it will change internal goldens.

Reply:
@LongCatIsLooong
A cleaner solution would be to change how the cursor height is computed.
The root of the problem is that the cursor height is based on `preferredLineHeight`.
Do you know why the cursor height is not computed in the same way as the selection highlight height?

Actually in Flutter we get this:

https://github.com/flutter/flutter/assets/840911/6afef77c-5446-4f63-b6d3-20c81e9cada1

Reply:
The selection highlight's height, with `EditableText`'s defaults, should roughly be the height of the font's EM box (so it does not scale with `TextStyle.height`). But On iOS it seems the cursor height is pretty close to the line height, and the height stays consistent within each line.

Reply:
Do you mean that cursor height being line height is WAI?
If so which solution can we try to mitigate the cursor height with M3 typography?